i6_models/parts/blstm.py

```diff
@@ -0,0 +1,76 @@
+from dataclasses import dataclass
+import torch
+from torch import nn
+
+from i6_models.config import ModelConfiguration
+
+
+@dataclass()
+class BlstmEncoderConfig(ModelConfiguration):
+    num_layers: int
+    """number of bi-directional LSTM layers, minimum 2"""
+    input_dimension: int
+    """input dimension size"""
+    hidden_dimension: int
+    """hidden dimension of one direction of LSTM, the total output size is twice of this"""
+    dropout: float
+    """nn.LSTM supports an internal Dropout layer"""
+    enforce_sorted: bool = True
+    """
+        keep activated for ONNX-Export, requires that the lengths are sorted decreasing from longest
+
+        Sorting can performed using something like:    
+
+            audio_features_len, indices = torch.sort(audio_features_len, descending=True)
+            audio_features = audio_features[indices, :, :]
+            labels = labels[indices, :]
+            labels_len = labels_len[indices]
+    """
+
+
+class BlstmEncoder(torch.nn.Module):
+    """
+    Simple multi-layer BLSTM model including dropout, batch-first variant,
+    hardcoded to use B,T,F input
+
+    supports: TorchScript, ONNX-export
+    """
+
+    def __init__(self, config: BlstmEncoderConfig):
+        """
+        :param config: configuration object
+        """
+        super().__init__()
+        self.dropout = config.dropout
+        self.enforce_sorted = config.enforce_sorted
+        self.blstm_stack = nn.LSTM(
+            input_size=config.input_dimension,
+            hidden_size=config.hidden_dimension,
+            bidirectional=True,
+            num_layers=config.num_layers,
+            batch_first=False,
+            dropout=self.dropout,
+        )
+
+    def forward(self, x: torch.Tensor, seq_len: torch.Tensor) -> torch.Tensor:
+        """
+        :param x: [B, T, input_dimension]
+        :param seq_len: [B], should be on CPU for Script/Trace mode
+        :return [B, T, 2 * hidden_dimension]
+        """
+        if not torch.jit.is_scripting() and not torch.jit.is_tracing():
+            # during graph mode we have to assume all Tensors are on the correct device,
+            # otherwise move lengths to the CPU if they are on GPU
+            if seq_len.get_device() >= 0:
+                seq_len = seq_len.cpu()
```

Comment:
Aren't we anyway always expecting this?

Reply:
In RETURNN yes, but I did not want to have that expectation here (I am also using the model in a different context myself)

Reply:
But this is somewhat relevant, if we maybe want to have that expectation always for all modules here in i6_models. 

That would simplify the code in many places, basically all places which involves seq lens, because as far as I understood, most (all?) typical code would always require seq lens on CPU. So if you don't have this expectation, you would need such code in many places.

And this can potentially be slow. If the seq len is really on GPU, it involves a GPU->CPU copy, which can be very slow.

Reply:
I have the seq_len always also on GPU, so that I can always assume all my tensors are on the target device. I also have setups that perform operations on the seq_len, e.g. within feature extraction or the conversion from seq_len to a mask, and I want to do this on GPU.

Still we can discuss to have the seq_len on CPU assumptions here in i6_models, I can find other solutions for me to deal with that.

Reply:
Was this discussed?

Reply:
Was this discussed now? What was the result of the discussion? I don't see this in the meeting notes.